### PR TITLE
handling errors from backend, adding new custom error pages

### PR DESCRIPTION
**Pull request by yangwong9664**

### CODE REVIEW

File: app/config/ErrorHandler.scala [hunk 1]
```diff
@@ -16,17 +16,33 @@
...
 @Singleton
-class ErrorHandler @Inject()(errorTemplate: ErrorTemplate, val messagesApi: MessagesApi)(implicit appConfig: AppConfig)
+class ErrorHandler @Inject()(errorTemplate: ErrorTemplate, val messagesApi: MessagesApi,
+                             internalServerErrorPage: InternalServerErrorPage, notFoundPage: NotFoundPage,
+                             serviceUnavailablePage: ServiceUnavailablePage)(implicit appConfig: FrontendAppConfig)
+
   extends FrontendErrorHandler {
 
   override def standardErrorTemplate(pageTitle: String, heading: String, message: String)(implicit request: Request[_]): Html =
     errorTemplate(pageTitle, heading, message)
```
Comment (yangwong9664) [inline app/config/ErrorHandler.scala:36]: replace with internal server error

File: app/connectors/httpparsers/IncomeSourcesHttpParser.scala [hunk 2]
```diff
@@ -16,36 +16,51 @@
...
+        case _ =>
+          pagerDutyLog(UNEXPECTED_RESPONSE_FROM_API, logMessage(response))
+          Left(IncomeSourcesUnhandledError)
       }
     }
-  }
-
 
+    private def logMessage(response:HttpResponse): Option[String] ={
+      Some(s"[SubmittedDividendsParser][read] Received ${response.status} from income-tax-dividends. Body:${response.body}")
```
Comment (yangwong9664) [inline app/connectors/httpparsers/IncomeSourcesHttpParser.scala:54]: change name

File: app/views/errors/InternalServerErrorPage.scala.html [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ *@
+
+@import uk.gov.hmrc.hmrcfrontend.views.html.components._
+
+@this(hmrcPageHeading : HmrcPageHeading)
+
+@()
+
+@hmrcPageHeading(PageHeading(
+text = "Sorry, there is a problem with the service"
```
Comment (yangwong9664) [inline app/views/errors/InternalServerErrorPage.scala.html:24]: put all text into messages file

File: app/views/errors/InternalServerErrorPage.scala.html [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+<p class="govuk-body">You can also:</p>
+<ul>
+ <li>
+  <p class="govuk-body">
+   go to the <a href="https://www.gov.uk/income-tax" class="govuk-link">Income Tax home page (opens in new tab)</a> for more information
+  </p>
+ </li>
+ <li>
+  <p class="govuk-body">
+   use <a href="https://www.gov.uk/government/organisations/hm-revenue-customs/contact/self-assessment" class="govuk-link">Self Assessment: general enquiries (opens in new tab)</a> to speak to someone about your income tax
```
Comment (yangwong9664) [inline app/views/errors/InternalServerErrorPage.scala.html:37]: fix new tab

File: app/views/errors/InternalServerErrorPage.scala.html [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+@import views.html.headerFooterTemplate.Layout
+
+@this(hmrcPageHeading: HmrcPageHeading, layout: Layout)
+
+@()(implicit request: Request[_], messages: Messages, appConfig: AppConfig)
+
+@layout(pageTitle = Some(messages("errors.500.heading"))) {
+
+@hmrcPageHeading(PageHeading(text = messages("errors.500.heading")))
+
```
Comment (robert-webb-hmrc) [inline app/views/errors/InternalServerErrorPage.scala.html:27]: Formatting of views needs addressing

**Review by robert-webb-hmrc — CHANGES_REQUESTED** [review]
So as it currently stands, we still need to address the NO_CONTENT response from the APIs. These are valid cases if the user is accessing the service with no prior submitted data.

**Comment by robert-webb-hmrc** [comment]
No Specs for the error views

File: app/config/ErrorHandler.scala [hunk 6]
```diff
@@ -17,16 +17,42 @@
...
+      case _ => InternalServerError(internalServerErrorPage())
+    }
+  }
+
+  override def onClientError(request: RequestHeader, statusCode: Int, message: String): Future[Result] =
+    statusCode match {
+      case play.mvc.Http.Status.NOT_FOUND =>
+        Future.successful(NotFound(notFoundTemplate(request.withBody(""))))
+      case _ =>
+        Future.successful(BadRequest(badRequestTemplate(request.withBody(""))))
```
Comment (BenCosfordHMRC) [inline app/config/ErrorHandler.scala:56]: sorry forgot about this one but could you make the case _ InternalServerError(internalServerErrorPage())

File: conf/messages [hunk 7]
```diff
@@ -52,4 +52,31 @@ overview.heading.isAgent.true = Your client’s Income Tax Return
...
+errors.500.p6 = use
+errors.500.p7 = Self Assessment: general enquiries (opens in new tab)
+errors.500.p8 = to speak to someone about your income tax
+
+errors.404.heading = Page not found
+errors.404.p1 = If you typed the web address, check it is correct.
+errors.404.p2 = If you used ‘copy and paste’ to enter the web address, check you copied the full address.
+errors.404.p3 = If the website address is correct or you selected a link or button, you can use
+errors.404.p4 = Self Assessment: general enquiries (opens in new tab)
+errors.404.p5 = to speak to someone about your income tax.
```
Comment (robert-webb-hmrc) [inline conf/messages:72]: This is inconsistent with the final line of the other error pages (full stop)

**Review by robert-webb-hmrc — CHANGES_REQUESTED** [review]
Few tidy up tasks to perform

File: conf/messages (hunk 7 above)
Comment (BenCosfordHMRC) [reply conf/messages:72]: This is okay as the other links are in bullet points (https://design-system.service.gov.uk/styles/typography/#bulleted-lists)

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/master/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Review by robert-webb-hmrc — APPROVED** [review]

**Pull request merged**